qiskit_machine_learning/kernels/quantum_kernel.py

```diff
@@ -0,0 +1,322 @@
+        otherwise construct :math:`Psi^dagger(y) x Psi(x)|0>`
+        If y is None and not using `statevector_simulator`, self inner product is calculated.
+
+        Args:
+            x:                  # first data point parameter vector
+            y:                  # second data point parameter vector,
+                                  ignored if using statevector simulator
+            measurement:        # include measurement if not using statevector simulator
+            is_statevector_sim: # use state vector simulator
+
+        Returns:
+            QuantumCircuit
+
+        Raises:
+            ValueError:
+                - x and/or y have incompatible dimension with feature map
+        """
+
+        if len(x) != self._feature_map.num_parameters:
+            raise ValueError("x and class feature map incompatible dimensions.\n" +
+                             "x has %s dimensions, but feature map has %s." %
+                             (len(x), self._feature_map.num_parameters))
+
+        q = QuantumRegister(self._feature_map.num_qubits, 'q')
+        c = ClassicalRegister(self._feature_map.num_qubits, 'c')
+        qc = QuantumCircuit(q, c)
+
+        x_dict = dict(zip(self._feature_map.ordered_parameters, x))
+        psi_x = self._feature_map.assign_parameters(x_dict)
+        qc.append(psi_x.to_instruction(), qc.qubits)
+
+        if not is_statevector_sim:
+            if y is not None and len(y) != self._feature_map.num_parameters:
+                raise ValueError("y and class feature map incompatible dimensions.\n" +
+                                 "y has %s dimensions, but feature map has %s." %
+                                 (len(y), self._feature_map.num_parameters))
+
+            if y is None:
+                y = x
+
+            y_dict = dict(zip(self._feature_map.ordered_parameters, y))
+            psi_y_dag = self._feature_map.assign_parameters(y_dict)
+            qc.append(psi_y_dag.to_instruction().inverse(), qc.qubits)
+
+            if measurement:
+                qc.barrier(q)
+                qc.measure(q, c)
+        return qc
+
+    def _compute_overlap(self, idx, results, is_statevector_sim, measurement_basis):
+        """
+        Helper function to compute overlap for given input.
+        """
+        if is_statevector_sim:
+            # |<0|Psi^dagger(y) x Psi(x)|0>|^2, take the amplitude
+            v_a, v_b = [results.get_statevector(int(i)) for i in idx]
+            tmp = np.vdot(v_a, v_b)
+            kernel_value = np.vdot(tmp, tmp).real  # pylint: disable=no-member
+        else:
+            result = results.get_counts(idx)
+
+            kernel_value = result.get(measurement_basis, 0) / sum(result.values())
+        return kernel_value
+
+    def evaluate(self, x_vec: np.ndarray, y_vec: np.ndarray = None) -> np.ndarray:
+        r"""
+        Construct kernel matrix for given data and feature map
+
+        If y is None, self inner product is calculated.
```

Comment:
If `y_vec` is None, would be more precise, 

Having said that are there better names that can be given to the params other than x_vec and y_vec which really don't say a lot. But then looking at kernel examples to sklearn they use X and Y so maybe things are fine if this is an understood use.

Reply:
Changed the comment to `y_vec`, but leaving the variable names as they are, since we want this to be easy for people familiar with sklearn to use.